feat(carto): Add 'tileResolution' and 'blockSize'

donmccurdy opened the pull request:
Adds user-facing `tileResolution` prop to CARTO layers, taking values `0.25 | 0.5 | 1 | 2 | 4`. Adds internal `blockSize` prop for raster layers, accepted from the backend API but not exposed as a user-facing prop.

- fixes #8587

donmccurdy commented:
Current dependencies on/for this PR:
* `master`
  * **PR #8502** <a href="https://app.graphite.dev/github/pr/visgl/deck.gl/8502?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/visgl/deck.gl/8502?utm_source=stack-comment).

felixpalmer commented on modules/carto/src/layers/schema/carto-raster-tile-loader.ts:
```diff
@@ -36,8 +36,12 @@ const CartoRasterTileLoader: LoaderWithParser = {
-  blockWidth: number;
-  blockHeight: number;
+  /** @deprecated Use blockSize. */
+  blockWidth?: number;
```
I don't think any public API was ever using these, I would just remove them to avoid confusion. When we deploy this we will bump the map version, so the client should only talk to a server that returns `blockSize`

felixpalmer commented on modules/carto/src/layers/raster-layer.ts:
```diff
@@ -31,7 +31,7 @@ class RasterColumnLayer extends ColumnLayer {
     const shaders = super.getShaders();
     const data = this.props.data as unknown as Raster & {length: number};
-    const BLOCK_WIDTH = data.blockWidth || Math.sqrt(data.length);
+    const BLOCK_WIDTH = data.blockSize || Math.sqrt(data.length);
```
Could you double check the logic here? We [have a bug](https://github.com/visgl/deck.gl/issues/8587) and I suspect that `Math.sqrt(data.length)` should in fact be `Math.sqrt(data.data.length)`

felixpalmer commented on modules/carto/src/layers/schema/carto-raster-tile-loader.ts:
```diff
@@ -50,13 +50,21 @@ function parseCartoRasterTile(
   if (!arrayBuffer) return null;
   const tile = parsePbf(arrayBuffer, TileReader);
-  const {bands, blockHeight, blockWidth} = tile;
+  const {bands, blockSize, blockWidth, blockHeight} = tile.bands;
```
```suggestion
  const {bands, blockSize, blockWidth, blockHeight} = tile;
```

felixpalmer commented on modules/carto/src/layers/schema/carto-raster-tile-loader.ts:
```diff
@@ -50,13 +50,21 @@ function parseCartoRasterTile(
     numericProps[name] = data;
   }
-  return {blockWidth, blockHeight, cells: {numericProps, properties: []}};
+  return {blockSize: blockSize || blockWidth, cells: {numericProps, properties: []}};
```
Just to be safe? Don't know why it would ever be 0, but maybe for an empty tile?
```suggestion
  return {blockSize: blockSize ?? blockWidth, cells: {numericProps, properties: []}};
```

felixpalmer commented on modules/carto/src/sources/types.ts:
```diff
@@ -75,6 +75,11 @@ export type QuerySourceOptions = {
   sqlQuery: string;
 
+  /**
+   * Relative resolution of a tile. Higher values increase density and data size. Default = 1.
```
As per discussion, for now this should remain 0.5

alasarr commented on modules/carto/src/sources/types.ts:
```diff
@@ -121,6 +126,11 @@ export type TableSourceOptions = {
   spatialDataColumn?: string;
+
+  /**
+   * Relative resolution of a tile. Higher values increase density and data size. Default = 1.
```
What about explaining it in more detail? 

When tileResoultion=`1` corresponds to a simplification where the coordinates within a tile a snapped to a 1024x1024 grid. `0.5` corresponds to a 512x512 grid, `4` 4096x4096 , etc...

donmccurdy replied:
Done!

donmccurdy commented:
I've broken a test that included a binary tile as a fixture, and will need to update that tile to the new schema:

https://github.com/visgl/deck.gl/blob/12b2532eff3cff5cfb12f148044ae9fcba16b372/test/modules/carto/layers/raster.spec.ts#L5-L7

I'm not sure what `tile 487624ffffffffff` refers to — unless someone else does, I'll pick a new raster tile for the test.

felixpalmer commented:
> I'm not sure what tile 487624ffffffffff refers

It's just some random tile I saved from the API, you can replace it

donmccurdy commented:
@felixpalmer I've updated the mock tile, and I believe also fixed the cause of #8587. If the map was initialized to a viewport without any raster tiles, the `BLOCK_WIDTH` parameter was initialized without a valid value, then shader compilation failed and never recovered. I suppose there's more than one way to fix that; I've defaulted block size to 0 to avoid the compile failure.
